### DISCUSSION TITLE
multi-thread ui

### DISCUSSION THREAD

**Discussion by AtomKing** [AvaloniaUI/Avalonia#10721] [Q&A]
I'm working on a paint application，when touch or mouse moved,i draw line between two points,but if there is too much strokes ,it will be very slow to render the draw canvas.
so ,i'm thinking ,if there 's  two or more UI thread,I can draw temporary data in a canvas  by other thread,when i lift finnger,i will clear all 
lines in this canvas ,but draw them in main UI thread
in wpf ,i use VisualTarget and HostVisual to achieve this function，i don't know how to do it with avalonia ,or other way to fix this problem

**Comment by kekekeks** [comment]
Input events are being grouped. See [PointerCanvas example from the control catalog](https://github.com/AvaloniaUI/Avalonia/blob/498f68f6597ffda2d3e90ab61c034fa12bc8f69b/samples/ControlCatalog/Pages/PointerCanvas.cs)

**Reply by AtomKing** [reply]
thanks for your replay,but it's not what i want ,this sample showed how to draw ,but same problem ,if you draw two much ,it will be very slow because the program must render all strokes when your touch or pen moved,but if i have two canvas ,when i move ,i draw canvas1,but not update canvas2,when touch up,clear canvas1,add stroke to canvas2,i thick  inkcanvas in wpf worked like this

**Reply by AtomKing** [reply]
writeablebitmap ?

**Reply by maxkatz6** [reply]
You can use writeablebitmap for caching, yes. Or render target bitmap to which you can draw using avalonia DrawLine apis.

**Reply by AtomKing** [reply]
i tried,but the same problem,avalonia has just one UI thread,if i need render sone uielement, all of them will be rendered again,it is not suitable for programs that refresh frequently .
but i realized，how mediaplayer did may be helpful

**Reply by timunie** [reply]
Check also the Render demo in the source code.
